ext/mixed/js/text-scanner.js

```diff
@@ -0,0 +1,386 @@
+            return;
+        }
+
+        this.preventScroll = false;
+        this.preventNextContextMenu = false;
+        this.preventNextMouseDown = false;
+        this.preventNextClick = false;
+
+        const primaryTouch = e.changedTouches[0];
+        if (DOM.isPointInSelection(primaryTouch.clientX, primaryTouch.clientY, this.node.getSelection())) {
+            return;
+        }
+
+        this.primaryTouchIdentifier = primaryTouch.identifier;
+
+        if (this.pendingLookup) {
+            return;
+        }
+
+        const textSourceCurrentPrevious = this.textSourceCurrent !== null ? this.textSourceCurrent.clone() : null;
+
+        this.searchAt(primaryTouch.clientX, primaryTouch.clientY, 'touchStart')
+        .then(() => {
+            if (
+                this.textSourceCurrent === null ||
+                this.textSourceCurrent.equals(textSourceCurrentPrevious)
+            ) {
+                return;
+            }
+
+            this.preventScroll = true;
+            this.preventNextContextMenu = true;
+            this.preventNextMouseDown = true;
+        });
+    }
+
+    onTouchEnd(e) {
+        if (
+            this.primaryTouchIdentifier === null ||
+            TextScanner.getIndexOfTouch(e.changedTouches, this.primaryTouchIdentifier) < 0
+        ) {
+            return;
+        }
+
+        this.primaryTouchIdentifier = null;
+        this.preventScroll = false;
+        this.preventNextClick = false;
+        // Don't revert context menu and mouse down prevention,
+        // since these events can occur after the touch has ended.
+        // this.preventNextContextMenu = false;
+        // this.preventNextMouseDown = false;
+    }
+
+    onTouchCancel(e) {
+        this.onTouchEnd(e);
+    }
+
+    onTouchMove(e) {
+        if (!this.preventScroll || !e.cancelable || this.primaryTouchIdentifier === null) {
+            return;
+        }
+
+        const touches = e.changedTouches;
+        const index = TextScanner.getIndexOfTouch(touches, this.primaryTouchIdentifier);
+        if (index < 0) {
+            return;
+        }
+
+        const primaryTouch = touches[index];
+        this.searchAt(primaryTouch.clientX, primaryTouch.clientY, 'touchMove');
+
+        e.preventDefault(); // Disable scroll
+    }
+
+    async onSearchClear() {
+        this.searchClear();
+        await this.publish('searchClear', {});
+    }
+
+    async onTextSearch(textSource, cause) {
+        this.pendingLookup = true;
+        const results = await this.publish('textSearch', {textSource, cause});
+        if (results.some((r) => r)) {
+            this.textSourceCurrent = textSource;
+            if (this.options.scanning.selectText) {
+                textSource.select();
+            }
+        }
+        this.pendingLookup = false;
+    }
+
+    onError(error) {
+        logError(error, false);
+    }
+
+    subscribe(eventName, subscriber) {
```

Comment:
Looks like an event listener pattern is implemented here with ```subscribe``` and ```publish```. Is there a situation where there will be multiple listeners for events? If not, you might be able to simplify the code by just having overridable stub functions, similar to either of the below:

https://github.com/FooSoft/yomichan/blob/14c5857dfe58eb14599438f8cfb8d436d38e2006/ext/mixed/js/display.js#L45-L47
https://github.com/FooSoft/yomichan/blob/14c5857dfe58eb14599438f8cfb8d436d38e2006/ext/bg/js/conditions-ui.js#L55-L57

The reason I bring this up is because typically the event listener pattern is synchronous, whereas this publish version is async.

It looks like there are two places where ```publish``` is called:
* ```await this.publish('searchClear', {});``` doesn't need to be async because the searchClear handler isn't async.
* ```await this.publish('textSearch', {textSource, cause});``` does need to be async, but it looks like it only expects one result.

I noticed this because I recently added a similar event listener pattern on one of my development branches (which isn't async), and also used the "overridable stub" pattern a bit.
https://github.com/toasted-nutbread/yomichan/commit/648f807e53adbfc3d7c33a8d8b26a022ea047842

Reply:
> Looks like an event listener pattern is implemented here with `subscribe` and `publish`.

Oops, you're right. I tried to choose a name that can't be confused with the already existing `eventListeners` array and code related to DOM event handling, but this implementation doesn't have an event bus.

> Is there a situation where there will be multiple listeners for events? If not, you might be able to simplify the code by just having overridable stub functions

I can't think of one for this use case or for query parser, but I usually like to use lists by default when writing generalized code (although this doesn't *have* to be generalized). I assume you meant that I would do `class Frontend extends TextScanner` and `class QueryParser extends TextScanner` in the future; one problem I see is that TextScanner needs access to the Popup element and its containsPoint method in `Frontend`, but because Popups don't exist in query parser, I don't want to make the code depend on it explicitly. Having a different namespace makes that distinction clearer. It could also be useful to be able to add a TextScanner to any number of elements.

Anyway I'm all for simplification where possible, and I'll see how it goes.

Reply:
It seemed to work quite nicely, and I managed to shave off some extra lines of code too. Thanks for the idea :+1: Are there other things you think could be better?

Reply:
> I assume you meant that I would do ...

There's a few ways of going about it:
* Inheritance
  Subclass and provide override definitions in the subclass. This is what you're currently doing after the update.
* Composition (pure)
  This is basically what you originally did: `Frontend` has a private reference to an instance of `TextScanner`, and it's implementation is not modified.
* Composition (hybrid)
  I'm not sure what the technical name of this would be, but this would be where `Frontend` has a private reference to `TextScanner`, and it overrides methods by assigning properties to the instance which override the prototype chain. This is kind of a mix of both.

Here are some examples of the "hybrid" composition pattern that have been used:
https://github.com/FooSoft/yomichan/blob/a5a6351cceacde6f440d28ff7f9afa7c6c1ff251/ext/bg/js/settings/audio.js#L25-L30

https://github.com/FooSoft/yomichan/blob/a5a6351cceacde6f440d28ff7f9afa7c6c1ff251/ext/bg/js/settings/profiles.js#L90-L100

I've generally found the third pattern to be the most brief of the three. Pick whichever one you feel is the best for this case. The only real deciding factor I can think of for this change would be whether you want `TextScanner`'s public APIs to be exposed on the `Frontend` class.

Reply:
I like your hybrid approach in that it makes it really obvious what's being overridden. Inheritance has always felt like magic to me, but at least it's well understood magic. It's also used to extend `Display` to get `DisplayFloat` and `DisplaySearch`, which is why I assumed you meant only that after seeing your example about `Display.onError`.

In this case the original code was already in a class, so extracting a superclass was probably the best first step. That way I can reuse the `hookEvents` architecture and don't have to duplicate a lot of code or do other refactoring.

I don't necessarily want `TextScanner`'s public APIs to be exposed, but right now `setTextSourceScanLength` is needed in `Frontend` and I think it's good that `ignoreNodes` and such are in `TextScanner` instead. Maybe `TextSourceElement` could could take `ignoreNodes` so that the method wouldn't be needed in textScanner.